### ISSUE TITLE
INSERT modifiers REPLACE and APPLY are broken

### ISSUE THREAD

**Issue by fillest** [ClickHouse/ClickHouse#60260]
Only EXCEPT modifier works for INSERT, but REPLACE and/or APPLY results in an error, probably due to incorrect parsing.

```sql
CREATE TABLE test (s String, t DateTime) Engine=Memory;

insert into test (s, * replace(toStartOfHour(t) as t) except(s) apply(max)) format values ('test', now());
```

> A link to reproducer in [https://fiddle.clickhouse.com/](https://fiddle.clickhouse.com/).

https://fiddle.clickhouse.com/b5053d9d-70e9-41a1-825d-44ed6aa29326

**Does it reproduce on the most recent release?**

Yes

**Error message and/or stacktrace**

```
Received exception from server (version 24.1.5):
Code: 16. DB::Exception: Received from localhost:9000. DB::Exception: No such column max(toStartOfHour(t)) in table default.test (e5087d88-b0a7-4b3f-a2bd-2a3547d720e5). (NO_SUCH_COLUMN_IN_TABLE)
(query: insert into test (s, * replace(toStartOfHour(t) as t) except(s) apply(max)) format values ('test', now());)
```

**Comment by alexey-milovidov** [comment]
@amosbird, PTAL.

**Comment by fillest** [comment]
Yes, my goal is to transform of the columns before/during insertion (to "truncate" a datetime with toStartOfHour()). The problem is that `insert ... select from input()` seems to be broken too in some ways - I get NUMBER_OF_COLUMNS_DOESNT_MATCH (I checked multiple times that the columns do match). Also it is much less convenient because that way I have to fully specify the schema, not just the name list. And it behaves in a weird way so I have to also specify the columns both in `select` and in `insert`, otherwise it throws an error about a missing column (which exists in the table but not in the file, but the column has a `default`). I do not have enough energy to report this currently with a reproducible description, sorry.

But I do not see how it actually matters here because [the INSERT docs](https://clickhouse.com/docs/en/sql-reference/statements/insert-into) state the following:
>You can also use an expression with column [matcher](https://clickhouse.com/docs/en/sql-reference/statements/select#asterisk) such as * and/or [modifiers](https://clickhouse.com/docs/en/sql-reference/statements/select#select-modifiers) such as [APPLY](https://clickhouse.com/docs/en/sql-reference/statements/select#apply-modifier), [EXCEPT](https://clickhouse.com/docs/en/sql-reference/statements/select#except-modifier), [REPLACE](https://clickhouse.com/docs/en/sql-reference/statements/select#replace-modifier).

This means that `APPLY`, `EXCEPT` and `REPLACE` should work for `insert` the same way they work for `select`. But they do not! Only `EXCEPT` works the same way but `APPLY` and `REPLACE` not only behave differently - it seems they currently can not be used in any way except as a no-op. You can not write e.g. `except replace(toStartOfHour(t) as t)`, you can only write `except replace(t as t)` which is a no-op. `No such column toStartOfHour(t) in tabl` makes no sense if `REPLACE` has the same meaning as for `select` - because with `select` there would be no `toStartOfHour(t)` column too, obviously. Because with `select` the column is `t` and `toStartOfHour(t)` is what gets computed and gets the alias `t`. 

So either the docs are wrong (should not state that `EXCEPT` and `REPLACE` can be used for `insert`) or incomplete (should state that `EXCEPT` and `REPLACE` behave differently or should be use somehow differently) - or `EXCEPT` and `REPLACE`  are implemented incorrectly for `insert`

**Comment by amosbird** [comment]
> The problem is that insert ... select from input() seems to be broken too in some ways - I get NUMBER_OF_COLUMNS_DOESNT_MATCH (I checked multiple times that the columns do match). Also it is much less convenient because that way I have to fully specify the schema, not just the name list

I can see that schema inference of `INSERT INTO SELECT FROM input()` is somehow limited and inconvenient. I think it's a valid feature request to extend `use_structure_from_insertion_table_in_table_functions` setting with a value `3`, which means to use the structure of INSERT table directly as inference. (as if explicitly write `input(<table_structure>)`)

> So either the docs are wrong (should not state that EXCEPT and REPLACE can be used for insert)

Yes it's wrong and needs to be fixed. Only `EXCEPT` modifier is meaningful to transform insert column list. (it's a column list, not an expression list).

BTW, your transformation alone is invalid: `max` is an aggregate function, but there exist other columns which are neither group by keys nor aggregates.

**Comment by fillest** [comment]
> `max` is an aggregate function

Yep, I used it just as an example to check how `APPLY` will be handled too. Initially I've tried only `REPLACE`.

>Yes it's wrong and needs to be fixed

OK.. Since you haven't reopened the issue I guess I should open a separate one - more specific about the docs then.
